src/main/java/newbank/client/ExampleClient.java

```diff
@@ -6,77 +6,102 @@
+  private final BufferedReader userInput;
+  private final Thread bankServerResponseThread;
   private Display display = new ConsoleDisplay();
 
+  /**
+   * @param ip an ip address, or the loopback address
+   * @param port the port that the newbank server will be listening on
+   * @throws ConfigurationException af the server cannot be reached at the supplied ip and port
+   * @throws IOException if an I/O exception occurs while trying to connect to the server
+   */
   public ExampleClient(String ip, int port, Reader reader)
-      throws UnknownHostException, IOException {
-    server = new Socket(ip, port);
+      throws IOException, IllegalThreadStateException, ConfigurationException {
+    server = getSocket(ip, port);
     userInput = new BufferedReader(reader);
     bankServerOut = new PrintWriter(server.getOutputStream(), true);
 
-    bankServerResponceThread =
+    bankServerResponseThread =
         new Thread() {
-          private BufferedReader bankServerIn =
+          private final BufferedReader bankServerIn =
               new BufferedReader(new InputStreamReader(server.getInputStream()));
 
           public void run() {
-            try {
-              while (true) {
-                String response = bankServerIn.readLine();
-
-                if (response == null) {
-                  // socket broken, end thread
-                  return;
-                }
-
-                display.writeLine(response);
+            while (true) {
+              String response = null;
+              try {
+                response = bankServerIn.readLine();
+              } catch (IOException e) {
+                e.printStackTrace();
+                Thread.currentThread().interrupt();
               }
-            } catch (IOException e) {
-              e.printStackTrace();
-              return;
+              if (response == null) break; // socket broken, end thread
+
+              display.writeLine(response);
             }
           }
         };
-    bankServerResponceThread.start();
+    bankServerResponseThread.start();
   }
 
   public void setDisplay(Display display) {
     this.display = display;
   }
 
-  public void run() {
-    while (true) {
-      try {
-        while (true) {
-          String command = userInput.readLine();
-
-          if (command == null) {
-            // EOF reached, end thread
-            return;
-          }
+  /**
+   * Tries to open a socket, but does some error-diagnosis if that fails
+   * @param ip an ip address, or the loopback address
+   * @param port the port that the newbank server will be listening on
+   * @return a socket connected to the newbank server
+   * @throws ConfigurationException af the server cannot be reached at the supplied ip and port
+   * @throws IOException if an I/O exception occurs while creating the socket
+   */
+  private Socket getSocket(String ip, int port) throws ConfigurationException, IOException {
+    try {
+      return new Socket(ip, port);
```

Comment:
This block of code needs re-viewing/discussions. The server is (in the real-world) running on a separate machine from the client. The client is not going to see the messages on display in the server.

Jules

Reply:
These errors are thrown in the client and displayed to the client machine.
See for yourself by running the client without the server.

Reply:
Correct. I think we should end up with a single catch handle that says what you have below, that is "The newbank server cannot be reached at ip %s, port number %d.", DEFAULT_IP, DEFAULT_PORT". The client is unlikely to change the port number, that is not a configurable parameter. The client here is expected to run the program as we give it to them.

Kindest Regards,
Jules

Reply:
Done. Since you insist ;)